internal/pkg/server/agent.go

```diff
@@ -472,3 +491,59 @@ func apmConfigToInstrumentation(src *proto.APMConfig) (config.Instrumentation, e
 	}
 	return config.Instrumentation{}, fmt.Errorf("unable to transform APMConfig to instrumentation")
 }
+
+// injectMissingOutputAttributes will insert any kv pairs that are present in bootstrap but are not in outMap.
+// If the value of an existing key in outMap and bootstrap is map[string]interface{} the same injection is called recursively.
+func injectMissingOutputAttributes(outMap, bootstrap map[string]interface{}) {
```

Comment:
Why do we need to do this?

Reply:
this will inject `service_token` and any tls or proxy settings into the output if they are not specified in the policy output

Reply:
If we can fall back to the bootstrap output, do we actually need to do this? Can't those things be configured in the Fleet UI? Having the user enter them again is better than auto-magically updating the outputs with parameters users can't see.

If Fleet Server always has the outputs defined in the policy, and the bootstrap output it was installed with, isn't there always a working output available? Or is there something fleet server specific that we don't allow configuring in the Fleet UI?

I am just double checking that we are not adding complexity that we don't need.

Reply:
I think there's a bit of a misunderstanding here.

Bootstrap config/output is whatever enrolment flags are used when elastic-agent is set to start a fleet-server instance (es host, tls options, proxy options, service token).
While these values are initially populated using the policy output they are not updated in the agent after they are used.

Currently the elastic-agent will overwrite any keys in the output block before sending the config to fleet-server.
This causes the agent to pass the singular host it was enrolled with instead of the hosts list that may be in the real policy output.
The PR changes it so that all bootstrap options appear under a new key and do not overwrite any existing values.
so the output hosts would be the real policy's list, and bootstrap.hosts is the one that agent was enrolled with.

Injection is used primarily to ensure the service_token value gets to fleet-server, as the fleet-server policy never contains the service_token.

Currently the bootstrap config is inserted with every config expected output unit change that elastic-agent sends to fleet-server as it was the simplest to implement/test.

Reply:
> Injection is used primarily to ensure the service_token value gets to fleet-server, as the fleet-server policy never contains the service_token.

Got it so we need to do this for some keys all of the time.

Can you reduce the scope of this function to inject only the things you know will always be missing? And log obviously that this has happened.

What I want us to avoid is writing a generic function that unconditionally injects any missing keys, regardless of if it is correct or desirable to do that. I am worried about us obsoleting a parameter, or having us inject a now invalid or expired TLS attribute, and breaking connectivity to ES inexplicibly.

I am a bit paranoid about generic configuration manipulation after watching the investigation that went into finding https://github.com/elastic/fleet-server/pull/3508 :)

Reply:
> A bad output option is added or there's a config change immediately followed by a network issue and fleet-server fails the (es) output check

> at this point the bootstrap values with the expired CA is used causing fleet-server to fail until the file is replaced and the agent is restarted.

The point about a transient network issue causing the output check to fail is interesting. This makes me think the output check needs to be periodic and not a one time event.

> For the 1st issue, i think we can alter the injection to remove verification_mode: none from being injected if the policy output has CAs or a fingerprint specified.

This makes sense but it will still seem a bit magical from the user's perspective, so we need to clearly document the behavior.

Both of the cases you listed above need to become tests.